tests/base/vectorization_make_vectorized_array_01.cc

```diff
@@ -0,0 +1,107 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2015 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// test make_vectorized_array function for all possible value_types and vector
+// lengths
+
+#include <deal.II/base/vectorization.h>
+
+#include <limits>
+
+#include "../tests.h"
+
+
+template <typename VectorizedArrayType>
+void
+do_test(const VectorizedArrayType                      array,
+        const typename VectorizedArrayType::value_type number)
+{
+  deallog << "  test " << VectorizedArrayType::n_array_elements
+          << " array elements" << std::endl;
+  for (unsigned int i = 0; i < VectorizedArrayType::n_array_elements; i++)
+    if (array[i] != number)
+      deallog << "  problem in element " << i << std::endl;
+}
+
+
+template <typename Number>
+struct Tester
+{
+  static void
+  test()
+  {}
+};
+
+template <>
+struct Tester<double>
+{
+  static void
+  test()
+  {
+    do_test(make_vectorized_array<double>(2.0), 2.0);
+
+#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+    do_test(make_vectorized_array<VectorizedArray<double, 8>>(2.0), 2.0);
+#endif
+
+#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 2 && defined(__AVX__)
+    do_test(make_vectorized_array<VectorizedArray<double, 4>>(2.0), 2.0);
+#endif
+
+#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1 && defined(__SSE2__)
+    do_test(make_vectorized_array<VectorizedArray<double, 2>>(2.0), 2.0);
+#endif
+
+    do_test(make_vectorized_array<VectorizedArray<double, 1>>(2.0), 2.0);
+  }
+};
+
+template <>
+struct Tester<float>
+{
+  static void
+  test()
+  {
+    do_test(make_vectorized_array<float>(2.0), 2.0);
```

Comment:
will this not now fail your static_assert?

Reply:
Why should it? There is still the old `make_vectorized_array<Number>`:
```cpp
template <
  typename Number,
  int width = internal::VectorizedArrayWidthSpecifier<Number>::max_width>
inline DEAL_II_ALWAYS_INLINE VectorizedArray<Number, width>
                             make_vectorized_array(const Number &u);
```
Or am missing something?

Reply:
Hm. How does the compiler distinguish between the two functions if you write `make_vectorized_array<float>(...)`?

Reply:
Does this probably get resolved due to SFINAE-like handling because one runs into the static assert whereas the other does compile fine?

Reply:
It's not because of the `static_assert`, but because you of using `VectorizedArrayType::value_type` in the signature, see https://godbolt.org/z/92xiyE.
This is a plain SFINAE case.

Reply:
Yes. Selecting a function is done purely by looking at its signature, not its implementation. But the `VectorizedArrayType::value_type` makes a difference. The compiler would complain about an ambiguity, though, if you said
```
  make_vectorized_array<std::vector<double>> (...);
```
I suspect we never want to create such a vectorized array, so that's ok. Otherwise, we could play with `std::enable_if`.

Reply:
I guess we can indeed live with the ambiguity for `VectorizedArray<ComplicatedType>` where `ComplicatedType` defines `value_type` such that the lookup would succeed. Since `VectorizedArray` is really only non-trivial for `double` and `float` types at this point, we should be pretty safe. As you say, we could fix it by `enable_if` in case we feel for it.